token-handler/src/mint_tokens.rs

```diff
@@ -1,58 +1,63 @@
-use multiversx_sc::api::ESDT_NFT_CREATE_FUNC_NAME;
-use multiversx_sc::types::ManagedArgBuffer;
-use multiversx_sc::types::{ManagedVec, MultiValueEncoded};
+use header_verifier::header_verifier_proxy;
+use multiversx_sc::api::{ESDT_MULTI_TRANSFER_FUNC_NAME, ESDT_NFT_CREATE_FUNC_NAME};
+use multiversx_sc::imports::IgnoreValue;
+use multiversx_sc::types::{
+    system_proxy, EsdtTokenPayment, ManagedArgBuffer, ManagedAsyncCallResult, ToSelf,
+};
+use multiversx_sc::types::{ManagedVec, TokenIdentifier};
 use multiversx_sc::{codec, err_msg};
-use transaction::OperationEsdtPayment;
+use transaction::{
+    GasLimit, OperationData, OperationEsdtPayment, OperationTuple, StolenFromFrameworkEsdtTokenData,
+};
 
-use crate::common;
+const CALLBACK_GAS: GasLimit = 10_000_000; // Increase if not enough
+const TRANSACTION_GAS: GasLimit = 30_000_000;
+
+use crate::{burn_tokens, common};
 
 #[multiversx_sc::module]
-pub trait MintTokens: utils::UtilsModule + common::storage::CommonStorage {
+pub trait MintTokensModule:
+    utils::UtilsModule
+    + common::storage::CommonStorage
+    + common::events::EventsModule
+    + burn_tokens::BurnTokensModule
+    + tx_batch_module::TxBatchModule
+{
     #[endpoint(mintTokens)]
-    fn mint_tokens(&self, operation_tokens: MultiValueEncoded<OperationEsdtPayment<Self::Api>>) {
+    fn mint_tokens(
+        &self,
+        hash_of_hashes: ManagedBuffer,
+        operation_tuple: OperationTuple<Self::Api>,
+    ) {
         let mut output_payments: ManagedVec<Self::Api, OperationEsdtPayment<Self::Api>> =
             ManagedVec::new();
 
-        for operation_token in operation_tokens {
+        for operation_token in operation_tuple.operation.tokens.iter() {
             let sov_prefix = self.sov_prefix().get();
+
             if !self.has_sov_prefix(&operation_token.token_identifier, &sov_prefix) {
                 output_payments.push(operation_token.clone());
                 continue;
             }
 
             let mut nonce = operation_token.token_nonce;
+
             if nonce == 0 {
-                self.send().esdt_local_mint(
+                self.tx()
+                    .to(ToSelf)
+                    .typed(system_proxy::UserBuiltinProxy)
+                    .esdt_local_mint(
+                        &operation_token.token_identifier,
+                        operation_token.token_nonce,
+                        &operation_token.token_data.amount,
+                    )
+                    .sync_call();
+            } else {
+                let arg_buffer = self.get_nft_create_args(
```

Comment:
Better to just make the function to `create_args` and `mint` and return the nonce